core/trino-main/src/main/java/io/trino/dispatcher/LocalDispatchQuery.java

```diff
@@ -137,25 +138,30 @@ private void waitForMinimumWorkers()
 
     private void startExecution(QueryExecution queryExecution)
     {
-        queryExecutor.execute(() -> {
-            if (stateMachine.transitionToDispatching()) {
-                try {
-                    querySubmitter.accept(queryExecution);
-                    if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
-                        queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
-                    }
+        if (stateMachine.transitionToDispatching()) {
+            try {
+                querySubmitter.accept(queryExecution);
+                if (notificationSentOrGuaranteed.compareAndSet(false, true)) {
+                    queryExecution.addFinalQueryInfoListener(queryMonitor::queryCompletedEvent);
                 }
-                catch (Throwable t) {
-                    // this should never happen but be safe
-                    stateMachine.transitionToFailed(t);
-                    log.error(t, "query submitter threw exception");
-                    throw t;
+
+                submitted.set(null);
+                try (SetThreadName ignored = new SetThreadName("Query-%s", queryExecution.getQueryId())) {
+                    queryExecution.start();
                 }
-                finally {
+            }
+            catch (Throwable t) {
+                // this should never happen but be safe
+                stateMachine.transitionToFailed(t);
+                log.error(t, "query submitter threw exception");
+                throw t;
+            }
+            finally {
+                if (!submitted.isDone()) {
```

Comment:
Why add the `if` statement? I believe that `set` on a done future is a noop.  Also the future could complete between the `isDone` check and `set`